### PR TITLE
[tune] Search alg checkpointing during training

### PR DESCRIPTION
**Pull request by richardliaw**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: doc/source/tune/api_docs/suggestion.rst [hunk 1]
```diff
@@ -72,6 +72,54 @@ Tune also provides helpful utilities to use with Search Algorithms:
...
+    experiment_1 = tune.run(
+        cost,
+        num_samples=5,
+        search_alg=search_alg,
+        verbose=0,
+        name="my-experiment-1",
+        local_dir="~/my_results")
+
+    search_alg2 = HyperOptSearch()
+    Searcher.restore_from_dir(
```
Comment (amogkam) [inline doc/source/tune/api_docs/suggestion.rst:119]: Why have two ways to do this? Either the search_alg can restore directly or through the Searcher.

Comment (amogkam) [reply doc/source/tune/api_docs/suggestion.rst:119]: It just complicates things IMO.

Comment (richardliaw) [reply doc/source/tune/api_docs/suggestion.rst:119]: I'm not sure what you're suggesting. Can you provide some code examples?

Comment (amogkam) [reply doc/source/tune/api_docs/suggestion.rst:119]: why not just `search_alg2.restore_from_dir(os.path.join("~/my_results/my-experiment-1"))`?

Comment (richardliaw) [reply doc/source/tune/api_docs/suggestion.rst:119]: > It just complicates things IMO.

Also, I'm happy to clarify the design decision, but I'm not sure what problems you're seeing nor the actionable item to take here. 

It'd be great if during review, the comments would clarify 1. the problem with the current solution and 2. if possible, propose something that addresses the state problems while meeting existing requirements.

Comment (amogkam) [reply doc/source/tune/api_docs/suggestion.rst:119]: I just wanted to understand the reasoning behind why there is 2 ways to do (what seems like) the same thing. To restore from a path we can do it directly from the search algorithm, but to restore from a directory we have to use a `Searcher`? Why not the same approach to both restore from paths and directories? I'm sure there's a reason for this but to me it's not clear in the docs.

Comment (richardliaw) [reply doc/source/tune/api_docs/suggestion.rst]: fixed!

File: python/ray/tune/suggest/suggestion.py [hunk 2]
```diff
@@ -294,6 +390,12 @@ def on_trial_complete(self, trial_id, result=None, error=False):
     def is_finished(self):
         return self._counter >= self._total_samples or self._finished
 
+    def save(self, checkpoint_path):
+        self.searcher.save(checkpoint_path)
+
+    def restore(self, checkpoint_path):
+        self.searcher.restore(cehckpoint_path)
```
Comment (krfricke) [inline python/ray/tune/suggest/suggestion.py:397]:
```suggestion
        self.searcher.restore(checkpoint_path)
```

**Review by krfricke — APPROVED** [review]
The code changes look good. I'm just confused by the concepts of `SearchAlgorithm` and `Searcher`, especially since the docs sometimes use them for each other (e.g. https://docs.ray.io/en/releases-0.8.4/tune-searchalg.html, calling a search algorithm object `searcher`).
But this is not in the scope of this PR.

**Pull request merged**